# Predicting sentiment from product reviews


The goal of this first notebook is to explore logistic regression and feature engineering with existing GraphLab functions.

In this notebook you will use product review data from Amazon.com to predict whether the sentiments about a product (from its reviews) are positive or negative.

* Use SFrames to do some feature engineering
* Train a logistic regression model to predict the sentiment of product reviews.
* Inspect the weights (coefficients) of a trained logistic regression model.
* Make a prediction (both class and probability) of sentiment for a new product review.
* Given the logistic regression weights, predictors and ground truth labels, write a function to compute the **accuracy** of the model.
* Inspect the coefficients of the logistic regression model and interpret their meanings.
* Compare multiple logistic regression models.

Let's get started!
    
## Fire up GraphLab Create

Make sure you have the latest version of GraphLab Create.

In [1]:
from __future__ import division
import graphlab
import math
import string

# Data preparation

We will use a dataset consisting of baby product reviews on Amazon.com.

In [2]:
products = graphlab.SFrame('amazon_baby.csv')

This non-commercial license of GraphLab Create for academic use is assigned to xinningh@gmail.com and will expire on October 24, 2017.


[INFO] graphlab.cython.cy_server: GraphLab Create v2.1 started. Logging: C:\Users\Martin\AppData\Local\Temp\graphlab_server_1482274542.log.0


Finished parsing file C:\Users\Martin\amazon_baby.csv

Parsing completed. Parsed 100 lines in 1.33821 secs.

------------------------------------------------------
Inferred types from first 100 line(s) of file as 
column_type_hints=[str,str,long]
If parsing fails due to incorrect types, you can correct
the inferred type list above and pass it to read_csv in
the column_type_hints argument
------------------------------------------------------


Read 105270 lines. Lines per second: 79950.7

Finished parsing file C:\Users\Martin\amazon_baby.csv

Parsing completed. Parsed 183531 lines in 1.61345 secs.

Now, let us see a preview of what the dataset looks like.

In [3]:
products

Columns:
	name	str
	review	str
	rating	int

Rows: 183531

Data:
+-------------------------------+-------------------------------+--------+
|              name             |             review            | rating |
+-------------------------------+-------------------------------+--------+
|    Planetwise Flannel Wipes   | These flannel wipes are OK... |   3    |
|     Planetwise Wipe Pouch     | it came early and was not ... |   5    |
| Annas Dream Full Quilt wit... | Very soft and comfortable ... |   5    |
| Stop Pacifier Sucking with... | This is a product well wor... |   5    |
| Stop Pacifier Sucking with... | All of my kids have cried ... |   5    |
| Stop Pacifier Sucking with... | When the Binky Fairy came ... |   5    |
| A Tale of Baby's Days with... | Lovely book, it's bound ti... |   4    |
| Baby Tracker&reg; - Daily ... | Perfect for new parents. W... |   5    |
| Baby Tracker&reg; - Daily ... | A friend of mine pinned th... |   5    |
| Baby Tracker&reg; - Daily ... | This has been an easy way ... |   4    |
+-------------------------------+-------------------------------+--------+
[183531 rows x 3 columns]
Note: Only the head of the SFrame is printed.
You can use print_rows(num_rows=m, num_columns=n) to print more rows and columns.

## Build the word count vector for each review

Let us explore a specific example of a baby product.


In [3]:
products[269]

{'name': 'The First Years Massaging Action Teether',
 'rating': 5L,
 'review': 'A favorite in our house!'}

Now, we will perform 2 simple data transformations:

1. Remove punctuation using [Python's built-in](https://docs.python.org/2/library/string.html) string functionality.
2. Transform the reviews into word-counts.

**Aside**. In this notebook, we remove all punctuations for the sake of simplicity. A smarter approach to punctuations would preserve phrases such as "I'd", "would've", "hadn't" and so forth. See [this page](https://www.cis.upenn.edu/~treebank/tokenization.html) for an example of smart handling of punctuations.

In [4]:
def remove_punctuation(text):
    import string
    return text.translate(None, string.punctuation) 

review_without_punctuation = products['review'].apply(remove_punctuation)
products['word_count'] = graphlab.text_analytics.count_words(review_without_punctuation)

Now, let us explore what the sample example above looks like after these 2 transformations. Here, each entry in the **word_count** column is a dictionary where the key is the word and the value is a count of the number of times the word occurs.

In [5]:
products[269]

{'name': 'The First Years Massaging Action Teether',
 'rating': 5L,
 'review': 'A favorite in our house!',
 'word_count': {'a': 1L, 'favorite': 1L, 'house': 1L, 'in': 1L, 'our': 1L}}

## Extract sentiments

We will **ignore** all reviews with *rating = 3*, since they tend to have a neutral sentiment.

In [6]:
products = products[products['rating'] != 3]
len(products)

166752

Now, we will assign reviews with a rating of 4 or higher to be *positive* reviews, while the ones with rating of 2 or lower are *negative*. For the sentiment column, we use +1 for the positive class label and -1 for the negative class label.

In [7]:
products['sentiment'] = products['rating'].apply(lambda rating : +1 if rating > 3 else -1)
products

Columns:
	name	str
	review	str
	rating	int
	word_count	dict
	sentiment	int

Rows: 166752

Data:
+-------------------------------+-------------------------------+--------+
|              name             |             review            | rating |
+-------------------------------+-------------------------------+--------+
|     Planetwise Wipe Pouch     | it came early and was not ... |   5    |
| Annas Dream Full Quilt wit... | Very soft and comfortable ... |   5    |
| Stop Pacifier Sucking with... | This is a product well wor... |   5    |
| Stop Pacifier Sucking with... | All of my kids have cried ... |   5    |
| Stop Pacifier Sucking with... | When the Binky Fairy came ... |   5    |
| A Tale of Baby's Days with... | Lovely book, it's bound ti... |   4    |
| Baby Tracker&reg; - Daily ... | Perfect for new parents. W... |   5    |
| Baby Tracker&reg; - Daily ... | A friend of mine pinned th... |   5    |
| Baby Tracker&reg; - Daily ... | This has been an easy way ... |   4    |
| Baby Tracker&reg; - Daily ... | I love this journal and ou... |   4    |
+-------------------------------+-------------------------------+--------+
+-------------------------------+-----------+
|           word_count          | sentiment |
+-------------------------------+-----------+
| {'and': 3L, 'love': 1L, 'i... |     1     |
| {'and': 2L, 'quilt': 1L, '... |     1     |
| {'and': 3L, 'ingenious': 1... |     1     |
| {'and': 2L, 'all': 2L, 'he... |     1     |
| {'and': 2L, 'cute': 1L, 'h... |     1     |
| {'shop': 1L, 'be': 1L, 'is... |     1     |
| {'and': 2L, 'all': 1L, 'ri... |     1     |
| {'and': 1L, 'fantastic': 1... |     1     |
| {'all': 1L, 'standarad': 1... |     1     |
| {'all': 2L, 'nannys': 1L, ... |     1     |
+-------------------------------+-----------+
[166752 rows x 5 columns]
Note: Only the head of the SFrame is printed.
You can use print_rows(num_rows=m, num_columns=n) to print more rows and columns.

Now, we can see that the dataset contains an extra column called **sentiment** which is either positive (+1) or negative (-1).

## Split data into training and test sets

Let's perform a train/test split with 80% of the data in the training set and 20% of the data in the test set. We use `seed=1` so that everyone gets the same result.

In [8]:
train_data, test_data = products.random_split(.8, seed=1)
print len(train_data)
print len(test_data)

133416
33336


# Train a sentiment classifier with logistic regression

We will now use logistic regression to create a sentiment classifier on the training data. This model will use the column **word_count** as a feature and the column **sentiment** as the target. We will use `validation_set=None` to obtain same results as everyone else.

**Note:** This line may take 1-2 minutes.

In [9]:
sentiment_model = graphlab.logistic_classifier.create(train_data,
                                                      target = 'sentiment',
                                                      features=['word_count'],
                                                      validation_set=None)

Logistic regression:

--------------------------------------------------------

Number of examples          : 133416

Number of classes           : 2

Number of feature columns   : 1

Number of unpacked features : 121712

Number of coefficients    : 121713

Starting L-BFGS

--------------------------------------------------------

+-----------+----------+-----------+--------------+-------------------+

| Iteration | Passes   | Step size | Elapsed Time | Training-accuracy |

+-----------+----------+-----------+--------------+-------------------+

| 1         | 5        | 0.000002  | 3.290967     | 0.840754          |

| 2         | 9        | 3.000000  | 6.896219     | 0.931350          |

| 3         | 10       | 3.000000  | 7.646899     | 0.882046          |

| 4         | 11       | 3.000000  | 8.387564     | 0.954076          |

| 5         | 12       | 3.000000  | 9.141243     | 0.960964          |

| 6         | 13       | 3.000000  | 9.898425     | 0.975033          |

+-----------+----------+-----------+--------------+-------------------+

TERMINATED: Terminated due to numerical difficulties.

This model may not be ideal. To improve it, consider doing one of the following:
(a) Increasing the regularization.
(b) Standardizing the input data.
(c) Removing highly correlated features.
(d) Removing `inf` and `NaN` values in the training data.

In [10]:
sentiment_model

Class                          : LogisticClassifier

Schema
------
Number of coefficients         : 121713
Number of examples             : 133416
Number of classes              : 2
Number of feature columns      : 1
Number of unpacked features    : 121712

Hyperparameters
---------------
L1 penalty                     : 0.0
L2 penalty                     : 0.01

Training Summary
----------------
Solver                         : lbfgs
Solver iterations              : 6
Solver status                  : TERMINATED: Terminated due to numerical difficulties.
Training time (sec)            : 10.6751

Settings
--------
Log-likelihood                 : inf

Highest Positive Coefficients
-----------------------------
word_count[mobileupdate]       : 41.9847
word_count[placeid]            : 41.7354
word_count[labelbox]           : 41.151
word_count[httpwwwamazoncomreviewrhgg6qp7tdnhbrefcmcrprcmtieutf8asinb00318cla0nodeid] : 40.0454
word_count[knobskeeping]       : 36.2091

Lowest Negative Coeff

**Aside**. You may get a warning to the effect of "Terminated due to numerical difficulties --- this model may not be ideal". It means that the quality metric (to be covered in Module 3) failed to improve in the last iteration of the run. The difficulty arises as the sentiment model puts too much weight on extremely rare words. A way to rectify this is to apply regularization, to be covered in Module 4. Regularization lessens the effect of extremely rare words. For the purpose of this assignment, however, please proceed with the model above.

Now that we have fitted the model, we can extract the weights (coefficients) as an SFrame as follows:

In [11]:
weights = sentiment_model.coefficients
weights.column_names()

['name', 'index', 'class', 'value', 'stderr']

In [12]:
print weights

+-------------+--------------+-------+------------------+--------+
|     name    |    index     | class |      value       | stderr |
+-------------+--------------+-------+------------------+--------+
| (intercept) |     None     |   1   |  1.30337080544   |  None  |
|  word_count |  recommend   |   1   |  0.303815600015  |  None  |
|  word_count |    highly    |   1   |  1.49183015276   |  None  |
|  word_count | disappointed |   1   |  -3.95748618393  |  None  |
|  word_count |     love     |   1   |  1.43301685439   |  None  |
|  word_count |      it      |   1   | 0.00986646490307 |  None  |
|  word_count |    planet    |   1   | -0.797764553926  |  None  |
|  word_count |     and      |   1   |  0.048449573172  |  None  |
|  word_count |     bags     |   1   |  0.165541436615  |  None  |
|  word_count |    wipes     |   1   | -0.0949937947269 |  None  |
+-------------+--------------+-------+------------------+--------+
[121713 rows x 5 columns]
Note: Only the head of the SFrame is

There are a total of `121713` coefficients in the model. Recall from the lecture that positive weights $w_j$ correspond to weights that cause positive sentiment, while negative weights correspond to negative sentiment. 

Fill in the following block of code to calculate how many *weights* are positive ( >= 0). (**Hint**: The `'value'` column in SFrame *weights* must be positive ( >= 0)).

In [13]:
print weights['value']

[1.3033708054360706, 0.3038156000152835, 1.4918301527571731, -3.957486183926777, 1.4330168543925341, 0.009866464903066394, -0.7977645539258387, 0.04844957317195975, 0.16554143661471182, -0.09499379472686426, -0.8171517163309098, 0.48841347880791114, 0.13137848076485506, 0.018252811627944388, -0.07690141552349346, -0.30259527729673896, 0.135472678412507, -0.05300047863793763, 0.03837878820789393, 0.16550664933686293, -0.030052358101251696, -0.24658014554003643, 7.39633708720406, 0.4263265257017666, 0.6715568214143195, 0.9743446387845555, 0.05293592927825717, 0.09287344889168749, 0.1300905285253519, 0.010833655305224267, 2.3923403419346965, -0.022727497767720723, 0.12190979941774738, 0.24942645220695933, 0.0404928422710866, 0.9085121049076831, 0.07785834042872437, 0.09671185203611958, 2.3923403419346965, 0.643215125747837, -0.2995159091753986, 0.11969744113002852, 0.24608706331761276, -0.004100889964292597, 0.5677491254714948, 0.8837698300028214, -0.10300729811977279, -1.6349556269710135

In [14]:
num_positive_weights = len(weights[weights['value'] >= 0])
num_negative_weights = len(weights[weights['value'] < 0])

print "Number of positive weights: %s " % num_positive_weights
print "Number of negative weights: %s " % num_negative_weights

Number of positive weights: 68419 
Number of negative weights: 53294 


**Quiz Question:** How many weights are >= 0?

## Making predictions with logistic regression

Now that a model is trained, we can make predictions on the **test data**. In this section, we will explore this in the context of 3 examples in the test dataset.  We refer to this set of 3 examples as the **sample_test_data**.

In [15]:
sample_test_data = test_data[10:13]
print sample_test_data['rating']
sample_test_data

[5L, 2L, 1L]


Columns:
	name	str
	review	str
	rating	int
	word_count	dict
	sentiment	int

Rows: 3

Data:
+-------------------------------+-------------------------------+--------+
|              name             |             review            | rating |
+-------------------------------+-------------------------------+--------+
|   Our Baby Girl Memory Book   | Absolutely love it and all... |   5    |
| Wall Decor Removable Decal... | Would not purchase again o... |   2    |
| New Style Trailing Cherry ... | Was so excited to get this... |   1    |
+-------------------------------+-------------------------------+--------+
+-------------------------------+-----------+
|           word_count          | sentiment |
+-------------------------------+-----------+
| {'and': 2L, 'all': 1L, 'lo... |     1     |
| {'and': 1L, 'wall': 1L, 't... |     -1    |
| {'all': 1L, 'money': 1L, '... |     -1    |
+-------------------------------+-----------+
[3 rows x 5 columns]

Let's dig deeper into the first row of the **sample_test_data**. Here's the full review:

In [16]:
sample_test_data[0]['review']

'Absolutely love it and all of the Scripture in it.  I purchased the Baby Boy version for my grandson when he was born and my daughter-in-law was thrilled to receive the same book again.'

That review seems pretty positive.

Now, let's see what the next row of the **sample_test_data** looks like. As we could guess from the sentiment (-1), the review is quite negative.

In [17]:
sample_test_data[1]['review']

'Would not purchase again or recommend. The decals were thick almost plastic like and were coming off the wall as I was applying them! The would NOT stick! Literally stayed stuck for about 5 minutes then started peeling off.'

We will now make a **class** prediction for the **sample_test_data**. The `sentiment_model` should predict **+1** if the sentiment is positive and **-1** if the sentiment is negative. Recall from the lecture that the **score** (sometimes called **margin**) for the logistic regression model  is defined as:

$$
\mbox{score}_i = \mathbf{w}^T h(\mathbf{x}_i)
$$ 

where $h(\mathbf{x}_i)$ represents the features for example $i$.  We will write some code to obtain the **scores** using GraphLab Create. For each row, the **score** (or margin) is a number in the range **[-inf, inf]**.

In [18]:
scores = sentiment_model.predict(sample_test_data)
print scores

[1L, -1L, -1L]


In [19]:
scores = sentiment_model.predict(sample_test_data, output_type='margin')
print scores

[6.734619727059322, -5.734130996760349, -14.668460404468654]


### Predicting sentiment

These scores can be used to make class predictions as follows:

$$
\hat{y} = 
\left\{
\begin{array}{ll}
      +1 & \mathbf{w}^T h(\mathbf{x}_i) > 0 \\
      -1 & \mathbf{w}^T h(\mathbf{x}_i) \leq 0 \\
\end{array} 
\right.
$$

Using scores, write code to calculate $\hat{y}$, the class predictions:

In [20]:
scores = scores.apply(lambda scores : +1 if scores > 0 else -1)
print scores

[1L, -1L, -1L]


Run the following code to verify that the class predictions obtained by your calculations are the same as that obtained from GraphLab Create.

In [21]:
print "Class predictions according to GraphLab Create:" 
print sentiment_model.predict(sample_test_data)

Class predictions according to GraphLab Create:
[1L, -1L, -1L]


**Checkpoint**: Make sure your class predictions match with the one obtained from GraphLab Create.

### Probability predictions

Recall from the lectures that we can also calculate the probability predictions from the scores using:
$$
P(y_i = +1 | \mathbf{x}_i,\mathbf{w}) = \frac{1}{1 + \exp(-\mathbf{w}^T h(\mathbf{x}_i))}.
$$

Using the variable **scores** calculated previously, write code to calculate the probability that a sentiment is positive using the above formula. For each row, the probabilities should be a number in the range **[0, 1]**.

In [22]:
import numpy as np
scores = sentiment_model.predict(sample_test_data, output_type='margin')
scores = np.array(scores)
prob = 1/(1+np.exp(-scores))
print prob


[  9.98812385e-01   3.22326818e-03   4.26155800e-07]


**Checkpoint**: Make sure your probability predictions match the ones obtained from GraphLab Create.

In [23]:
print "Class predictions according to GraphLab Create:" 
print sentiment_model.predict(sample_test_data, output_type='probability')

Class predictions according to GraphLab Create:
[0.9988123848377198, 0.0032232681818003847, 4.2615579966548425e-07]


** Quiz Question:** Of the three data points in **sample_test_data**, which one (first, second, or third) has the **lowest probability** of being classified as a positive review?

# Find the most positive (and negative) review

We now turn to examining the full test dataset, **test_data**, and use GraphLab Create to form predictions on all of the test data points for faster performance.

Using the `sentiment_model`, find the 20 reviews in the entire **test_data** with the **highest probability** of being classified as a **positive review**. We refer to these as the "most positive reviews."

To calculate these top-20 reviews, use the following steps:
1.  Make probability predictions on **test_data** using the `sentiment_model`. (**Hint:** When you call `.predict` to make predictions on the test data, use option `output_type='probability'` to output the probability rather than just the most likely class.)
2.  Sort the data according to those predictions and pick the top 20. (**Hint:** You can use the `.topk` method on an SFrame to find the top k rows sorted according to the value of a specified column.)

In [24]:
test_data['prob_predict'] = sentiment_model.predict(test_data, output_type='probability')
print test_data

+-------------------------------+-------------------------------+--------+
|              name             |             review            | rating |
+-------------------------------+-------------------------------+--------+
| Baby Tracker&reg; - Daily ... | This has been an easy way ... |   4    |
| Baby Tracker&reg; - Daily ... | I love this journal and ou... |   4    |
| Nature's Lullabies First Y... | I love this little calende... |   5    |
| Nature's Lullabies Second ... | I had a hard time finding ... |   5    |
|  Lamaze Peekaboo, I Love You  | One of baby's first and fa... |   4    |
|  Lamaze Peekaboo, I Love You  | My son loved this book as ... |   5    |
|  Lamaze Peekaboo, I Love You  | Our baby loves this book &... |   5    |
| SoftPlay Giggle Jiggle Fun... | This bear is absolutely ad... |   2    |
| SoftPlay Peek-A-Boo Where'... | I bought two for recent ba... |   5    |
| Baby's First Year Undated ... | I searched high and low fo... |   5    |
+------------------------

In [25]:
test_data.topk('prob_predict', 20)

Columns:
	name	str
	review	str
	rating	int
	word_count	dict
	sentiment	int
	prob_predict	float

Rows: 20

Data:
+-------------------------------+-------------------------------+--------+
|              name             |             review            | rating |
+-------------------------------+-------------------------------+--------+
| Britax Decathlon Convertib... | I researched a few differe... |   4    |
| Ameda Purely Yours Breast ... | As with many new moms, I c... |   4    |
| Traveling Toddler Car Seat... | I am sure this product wor... |   2    |
| Shermag Glider Rocker Comb... | After searching in stores ... |   4    |
| Cloud b Sound Machine Soot... | First off, I love plush sh... |   5    |
| JP Lizzy Chocolate Ice Cla... | I got this bag as a presen... |   4    |
| Fisher-Price Rainforest Me... | My daughter wasn't able to... |   5    |
| Lilly Gold Sit 'n' Stroll ... | I just completed a two-mon... |   5    |
|  Fisher-Price Deluxe Jumperoo | I had already decided that... |   5    |
| North States Supergate Pre... | I got a couple of these to... |   4    |
+-------------------------------+-------------------------------+--------+
+-------------------------------+-----------+--------------+
|           word_count          | sentiment | prob_predict |
+-------------------------------+-----------+--------------+
| {'all': 1L, 'toted': 1L, '... |     1     |     1.0      |
| {'lansinoh': 1L, 'six': 1L... |     1     |     1.0      |
| {'all': 1L, 'just': 1L, 'p... |     -1    |     1.0      |
| {'spotclean': 1L, 'code': ... |     1     |     1.0      |
| {'cute': 1L, 'all': 2L, 's... |     1     |     1.0      |
| {'knot': 1L, 'just': 4L, '... |     1     |     1.0      |
| {'cute': 1L, 'all': 3L, 's... |     1     |     1.0      |
| {'son': 2L, 'infant': 1L, ... |     1     |     1.0      |
| {'all': 1L, 'just': 1L, 's... |     1     |     1.0      |
| {'gates': 5L, 'just': 1L, ... |     1     |     1.0      |
+-------------------------------+-----------+--------------+
[20 rows x 6 columns]
Note: Only the head of the SFrame is printed.
You can use print_rows(num_rows=m, num_columns=n) to print more rows and columns.

In [26]:
test_data.sort('prob_predict', ascending = False)[0:20]

Columns:
	name	str
	review	str
	rating	int
	word_count	dict
	sentiment	int
	prob_predict	float

Rows: 20

Data:
+-------------------------------+-------------------------------+--------+
|              name             |             review            | rating |
+-------------------------------+-------------------------------+--------+
| KidCo Magnet Lock Starter Set | If you're reading this on ... |   4    |
| Fisher-Price Rainforest Me... | My daughter wasn't able to... |   5    |
| Baby Jogger City Mini GT S... | Amazing, Love, Love, Love ... |   5    |
| Eddie Bauer Trailmaker Tra... | What I like about my trave... |   5    |
| Baby Jogger City Mini GT D... | We are well pleased with t... |   4    |
| Fisher-Price Grow with Me ... | Simple design in both body... |   5    |
| Maxi-Cosi Pria 70 with Tin... | We love this car seat!! It... |   5    |
| Orbit Baby Stroller Travel... | My daughter was born a pre... |   5    |
| Regalo Easy Step Walk Thru... | These gates so easy to ins... |   5    |
| Baby BeeHinds Magic Alls M... | My baby girl really loves ... |   4    |
+-------------------------------+-------------------------------+--------+
+-------------------------------+-----------+--------------+
|           word_count          | sentiment | prob_predict |
+-------------------------------+-----------+--------------+
| {'help': 1L, 'being': 1L, ... |     1     |     1.0      |
| {'cute': 1L, 'all': 3L, 's... |     1     |     1.0      |
| {'all': 3L, 'just': 1L, 'o... |     1     |     1.0      |
| {'all': 2L, 'pound': 2L, '... |     1     |     1.0      |
| {'infant': 1L, 'drasticall... |     1     |     1.0      |
| {'rounded': 1L, 'just': 1L... |     1     |     1.0      |
| {'all': 2L, 'coach': 2L, '... |     1     |     1.0      |
| {'rocker': 1L, 'all': 3L, ... |     1     |     1.0      |
| {'and': 5L, 'gates': 2L, '... |     1     |     1.0      |
| {'cdd': 1L, 'all': 1L, 'se... |     1     |     1.0      |
+-------------------------------+-----------+--------------+
[20 rows x 6 columns]
Note: Only the head of the SFrame is printed.
You can use print_rows(num_rows=m, num_columns=n) to print more rows and columns.

**Quiz Question**: Which of the following products are represented in the 20 most positive reviews? [multiple choice]


Now, let us repeat this exercise to find the "most negative reviews." Use the prediction probabilities to find the  20 reviews in the **test_data** with the **lowest probability** of being classified as a **positive review**. Repeat the same steps above but make sure you **sort in the opposite order**.

In [27]:
test_data.sort('prob_predict')[0:20]

Columns:
	name	str
	review	str
	rating	int
	word_count	dict
	sentiment	int
	prob_predict	float

Rows: 20

Data:
+-------------------------------+-------------------------------+--------+
|              name             |             review            | rating |
+-------------------------------+-------------------------------+--------+
| Jolly Jumper Arctic Sneak ... | I am a "research-aholic" i... |   5    |
| Levana Safe N'See Digital ... | This is the first review I... |   1    |
| Snuza Portable Baby Moveme... | I would have given the pro... |   1    |
| Fisher-Price Ocean Wonders... | We have not had ANY luck w... |   2    |
| VTech Communications Safe ... | This is my second video mo... |   1    |
| Safety 1st High-Def Digita... | We bought this baby monito... |   1    |
| Chicco Cortina KeyFit 30 T... | My wife and I have used th... |   1    |
| Prince Lionheart Warmies W... | *****IMPORTANT UPDATE*****... |   1    |
| Valco Baby Tri-mode Twin S... | I give one star to the dim... |   1    |
| Adiri BPA Free Natural Nur... | I will try to write an obj... |   2    |
+-------------------------------+-------------------------------+--------+
+-------------------------------+-----------+--------------------+
|           word_count          | sentiment |    prob_predict    |
+-------------------------------+-----------+--------------------+
| {'raining': 1L, 'all': 8L,... |     1     | 7.80415068208e-100 |
| {'all': 2L, 'just': 1L, 'b... |     -1    | 6.83650885513e-25  |
| {'contacted': 1L, 'being':... |     -1    | 2.12654510823e-24  |
| {'fishstarfish': 1L, 'all'... |     -1    | 2.24582080779e-23  |
| {'all': 4L, 'reviewers': 1... |     -1    | 1.32962966148e-22  |
| {'all': 3L, 'being': 1L, '... |     -1    | 2.06872097469e-20  |
| {'all': 4L, 'wrestle': 1L,... |     -1    | 5.93881994671e-20  |
| {'less': 1L, 'move': 1L, '... |     -1    | 6.28510016538e-20  |
| {'limited': 2L, 'forget': ... |     -1    | 8.05528712689e-20  |
| {'all': 2L, 'forget': 1L, ... |     -1    | 8.46521724941e-20  |
+-------------------------------+-----------+--------------------+
[20 rows x 6 columns]
Note: Only the head of the SFrame is printed.
You can use print_rows(num_rows=m, num_columns=n) to print more rows and columns.

**Quiz Question**: Which of the following products are represented in the 20 most negative reviews?  [multiple choice]

## Compute accuracy of the classifier

We will now evaluate the accuracy of the trained classifier. Recall that the accuracy is given by


$$
\mbox{accuracy} = \frac{\mbox{# correctly classified examples}}{\mbox{# total examples}}
$$

This can be computed as follows:

* **Step 1:** Use the trained model to compute class predictions (**Hint:** Use the `predict` method)
* **Step 2:** Count the number of data points when the predicted class labels match the ground truth labels (called `true_labels` below).
* **Step 3:** Divide the total number of correct predictions by the total number of data points in the dataset.

Complete the function below to compute the classification accuracy:

In [28]:
def get_classification_accuracy(model, data, true_labels):
    # First get the predictions
    ## YOUR CODE HERE
    predictions = model.predict(data)
    
    # Compute the number of correctly classified examples
    ## YOUR CODE HERE
    num_correct = 0
    for i in xrange(len(predictions)):
        if predictions[i] == true_labels[i]:
            num_correct = num_correct + 1

    # Then compute accuracy by dividing num_correct by total number of examples
    ## YOUR CODE HERE
    accuracy = num_correct / len(predictions)
    
    return accuracy

Now, let's compute the classification accuracy of the **sentiment_model** on the **test_data**.

In [29]:
get_classification_accuracy(sentiment_model, test_data, test_data['sentiment'])

0.9145368370530358

**Quiz Question**: What is the accuracy of the **sentiment_model** on the **test_data**? Round your answer to 2 decimal places (e.g. 0.76).

**Quiz Question**: Does a higher accuracy value on the **training_data** always imply that the classifier is better?

## Learn another classifier with fewer words

There were a lot of words in the model we trained above. We will now train a simpler logistic regression model using only a subset of words that occur in the reviews. For this assignment, we selected a 20 words to work with. These are:

In [30]:
significant_words = ['love', 'great', 'easy', 'old', 'little', 'perfect', 'loves', 
      'well', 'able', 'car', 'broke', 'less', 'even', 'waste', 'disappointed', 
      'work', 'product', 'money', 'would', 'return']

In [31]:
len(significant_words)

20

For each review, we will use the **word_count** column and trim out all words that are **not** in the **significant_words** list above. We will use the [SArray dictionary trim by keys functionality]( https://dato.com/products/create/docs/generated/graphlab.SArray.dict_trim_by_keys.html). Note that we are performing this on both the training and test set.

In [32]:
train_data['word_count_subset'] = train_data['word_count'].dict_trim_by_keys(significant_words, exclude=False)
test_data['word_count_subset'] = test_data['word_count'].dict_trim_by_keys(significant_words, exclude=False)

Let's see what the first example of the dataset looks like:

In [33]:
train_data[0]['review']

'it came early and was not disappointed. i love planet wise bags and now my wipe holder. it keps my osocozy wipes moist and does not leak. highly recommend it.'

The **word_count** column had been working with before looks like the following:

In [34]:
print train_data[0]['word_count']

{'and': 3L, 'love': 1L, 'it': 3L, 'highly': 1L, 'osocozy': 1L, 'bags': 1L, 'leak': 1L, 'moist': 1L, 'does': 1L, 'recommend': 1L, 'was': 1L, 'wipes': 1L, 'disappointed': 1L, 'early': 1L, 'not': 2L, 'now': 1L, 'holder': 1L, 'wipe': 1L, 'keps': 1L, 'wise': 1L, 'i': 1L, 'planet': 1L, 'my': 2L, 'came': 1L}


Since we are only working with a subset of these words, the column **word_count_subset** is a subset of the above dictionary. In this example, only 2 `significant words` are present in this review.

In [35]:
print train_data[0]['word_count_subset']

{'love': 1L, 'disappointed': 1L}


## Train a logistic regression model on a subset of data

We will now build a classifier with **word_count_subset** as the feature and **sentiment** as the target. 

In [36]:
simple_model = graphlab.logistic_classifier.create(train_data,
                                                   target = 'sentiment',
                                                   features=['word_count_subset'],
                                                   validation_set=None)
simple_model

Logistic regression:

--------------------------------------------------------

Number of examples          : 133416

Number of classes           : 2

Number of feature columns   : 1

Number of unpacked features : 20

Number of coefficients    : 21

Starting Newton Method

--------------------------------------------------------

+-----------+----------+--------------+-------------------+

| Iteration | Passes   | Elapsed Time | Training-accuracy |

+-----------+----------+--------------+-------------------+

| 1         | 2        | 0.248223     | 0.862917          |

| 2         | 3        | 0.394355     | 0.865713          |

| 3         | 4        | 0.543490     | 0.866478          |

| 4         | 5        | 0.678111     | 0.866748          |

| 5         | 6        | 0.803725     | 0.866815          |

| 6         | 7        | 0.940348     | 0.866815          |

+-----------+----------+--------------+-------------------+

SUCCESS: Optimal solution found.

Class                          : LogisticClassifier

Schema
------
Number of coefficients         : 21
Number of examples             : 133416
Number of classes              : 2
Number of feature columns      : 1
Number of unpacked features    : 20

Hyperparameters
---------------
L1 penalty                     : 0.0
L2 penalty                     : 0.01

Training Summary
----------------
Solver                         : newton
Solver iterations              : 6
Solver status                  : SUCCESS: Optimal solution found.
Training time (sec)            : 0.9809

Settings
--------
Log-likelihood                 : 44323.7254

Highest Positive Coefficients
-----------------------------
word_count_subset[loves]       : 1.6773
word_count_subset[perfect]     : 1.5145
word_count_subset[love]        : 1.3654
(intercept)                    : 1.2995
word_count_subset[easy]        : 1.1937

Lowest Negative Coefficients
----------------------------
word_count_subset[disappointed] : -2.3551
wo

We can compute the classification accuracy using the `get_classification_accuracy` function you implemented earlier.

In [37]:
get_classification_accuracy(simple_model, test_data, test_data['sentiment'])

0.8693004559635229

Now, we will inspect the weights (coefficients) of the **simple_model**:

In [38]:
simple_model.coefficients

Columns:
	name	str
	index	str
	class	int
	value	float
	stderr	float

Rows: 21

Data:
+-------------------+--------------+-------+-----------------+-----------------+
|        name       |    index     | class |      value      |      stderr     |
+-------------------+--------------+-------+-----------------+-----------------+
|    (intercept)    |     None     |   1   |   1.2995449552  | 0.0120888541331 |
| word_count_subset | disappointed |   1   |  -2.35509250061 | 0.0504149888557 |
| word_count_subset |     love     |   1   |  1.36543549368  | 0.0303546295109 |
| word_count_subset |     well     |   1   |  0.504256746398 |  0.021381300631 |
| word_count_subset |   product    |   1   | -0.320555492996 | 0.0154311321362 |
| word_count_subset |    loves     |   1   |  1.67727145556  | 0.0482328275384 |
| word_count_subset |    little    |   1   |  0.520628636025 | 0.0214691475665 |
| word_count_subset |     work     |   1   | -0.621700012425 | 0.0230330597946 |
| word_count_subset |     easy     |   1   |  1.19366189833  |  0.029288869202 |
| word_count_subset |    great     |   1   |  0.94469126948  | 0.0209509926591 |
+-------------------+--------------+-------+-----------------+-----------------+
[21 rows x 5 columns]
Note: Only the head of the SFrame is printed.
You can use print_rows(num_rows=m, num_columns=n) to print more rows and columns.

Let's sort the coefficients (in descending order) by the **value** to obtain the coefficients with the most positive effect on the sentiment.

In [39]:
simple_model.coefficients.sort('value', ascending=False).print_rows(num_rows=21)

+-------------------+--------------+-------+-----------------+-----------------+
|        name       |    index     | class |      value      |      stderr     |
+-------------------+--------------+-------+-----------------+-----------------+
| word_count_subset |    loves     |   1   |  1.67727145556  | 0.0482328275384 |
| word_count_subset |   perfect    |   1   |  1.51448626703  |  0.049861952294 |
| word_count_subset |     love     |   1   |  1.36543549368  | 0.0303546295109 |
|    (intercept)    |     None     |   1   |   1.2995449552  | 0.0120888541331 |
| word_count_subset |     easy     |   1   |  1.19366189833  |  0.029288869202 |
| word_count_subset |    great     |   1   |  0.94469126948  | 0.0209509926591 |
| word_count_subset |    little    |   1   |  0.520628636025 | 0.0214691475665 |
| word_count_subset |     well     |   1   |  0.504256746398 |  0.021381300631 |
| word_count_subset |     able     |   1   |  0.191438302295 | 0.0337581955697 |
| word_count_subset |     ol

**Quiz Question**: Consider the coefficients of **simple_model**. There should be 21 of them, an intercept term + one for each word in **significant_words**. How many of the 20 coefficients (corresponding to the 20 **significant_words** and *excluding the intercept term*) are positive for the `simple_model`?

In [ ]:
10

**Quiz Question**: Are the positive words in the **simple_model** (let us call them `positive_significant_words`) also positive words in the **sentiment_model**?

In [40]:
significant_words = ['love', 'great', 'easy', 'old', 'little', 'perfect', 'loves', 
      'well', 'able', 'car', 'broke', 'less', 'even', 'waste', 'disappointed', 
      'work', 'product', 'money', 'would', 'return']
table = sentiment_model.coefficients
table.filter_by(significant_words,'index').print_rows(num_rows=20)


+------------+--------------+-------+------------------+--------+
|    name    |    index     | class |      value       | stderr |
+------------+--------------+-------+------------------+--------+
| word_count | disappointed |   1   |  -3.95748618393  |  None  |
| word_count |     love     |   1   |  1.43301685439   |  None  |
| word_count |     well     |   1   |  0.627964877567  |  None  |
| word_count |   product    |   1   | -0.0430899241758 |  None  |
| word_count |    loves     |   1   |   1.5664851757   |  None  |
| word_count |    little    |   1   |  0.674162457499  |  None  |
| word_count |     work     |   1   | -0.605153835963  |  None  |
| word_count |     easy     |   1   |  1.21346937822   |  None  |
| word_count |    great     |   1   |  1.31459245039   |  None  |
| word_count |    would     |   1   | -0.360326995525  |  None  |
| word_count |     able     |   1   |  0.174331272552  |  None  |
| word_count |   perfect    |   1   |  1.75190114392   |  None  |
| word_cou

# Comparing models

We will now compare the accuracy of the **sentiment_model** and the **simple_model** using the `get_classification_accuracy` method you implemented above.

First, compute the classification accuracy of the **sentiment_model** on the **train_data**:

In [41]:
get_classification_accuracy(sentiment_model, train_data, train_data['sentiment'])

0.979440247046831

Now, compute the classification accuracy of the **simple_model** on the **train_data**:

In [42]:
get_classification_accuracy(simple_model, train_data, train_data['sentiment'])

0.8668150746537147

**Quiz Question**: Which model (**sentiment_model** or **simple_model**) has higher accuracy on the TRAINING set?

Now, we will repeat this exercise on the **test_data**. Start by computing the classification accuracy of the **sentiment_model** on the **test_data**:

In [43]:
get_classification_accuracy(sentiment_model, test_data, test_data['sentiment'])

0.9145368370530358

Next, we will compute the classification accuracy of the **simple_model** on the **test_data**:

In [44]:
get_classification_accuracy(simple_model, test_data, test_data['sentiment'])

0.8693004559635229

**Quiz Question**: Which model (**sentiment_model** or **simple_model**) has higher accuracy on the TEST set?

## Baseline: Majority class prediction

It is quite common to use the **majority class classifier** as the a baseline (or reference) model for comparison with your classifier model. The majority classifier model predicts the majority class for all data points. At the very least, you should healthily beat the majority class classifier, otherwise, the model is (usually) pointless.

What is the majority class in the **train_data**?

In [45]:
num_positive  = (train_data['sentiment'] == +1).sum()
num_negative = (train_data['sentiment'] == -1).sum()
print num_positive
print num_negative

112164
21252


Now compute the accuracy of the majority class classifier on **test_data**.

**Quiz Question**: Enter the accuracy of the majority class classifier model on the **test_data**. Round your answer to two decimal places (e.g. 0.76).

In [46]:
num_positive  = (test_data['sentiment'] == +1).sum()
num_negative = (test_data['sentiment'] == -1).sum()
total = num_positive + num_negative
accuracy_majority = num_positive / total
print accuracy_majority


0.842782577394


**Quiz Question**: Is the **sentiment_model** definitely better than the majority class classifier (the baseline)?